src/js/trove/particle.js

```diff
@@ -0,0 +1,154 @@
+
+  return util.definePyretModule(
+    "particle",
+    [],
+    {
+      values: {
+      },
+      aliases: {
+      },
+      datatypes: {
+      }
+    },
+    function(runtime, namespace) {
+      return runtime.loadJSModules(namespace, [worldLib, ffiLib], function(rawJsworld, ffi) {
+        return runtime.loadModulesNew(namespace, [world, pShimStruct], function(pWorld, pStruct) {
+          var WorldConfigOption = runtime.getField(pWorld, "internal").WorldConfigOption;
+          var adaptWorldFunction = runtime.getField(pWorld, "internal").adaptWorldFunction;
+
+          var sd_to_js = function(sd) {
+            var ret = {};
+            var arr = ffi.toArray(runtime.getField(sd, "keys-list").app());
+            for(var i in arr) {
+              var k = arr[i];
+              var v = runtime.getField(sd, "get-value").app(k);
+              if(typeof(v) === "string") {
+                ret[k] = v;
+              } else {
+                throw new Error('unimplemented value conversion for StringDict');
+              }
+            }
+            return ret;
+          };
+
+          var OnParticle = function(handler, name, sdict) {
+            WorldConfigOption.call(this, 'on-particle');
+            this.handler = handler;
+            this.event = name;
+            this.options = sd_to_js(sdict);
+          };
+
+          OnParticle.prototype = Object.create(WorldConfigOption.prototype);
+
+          OnParticle.prototype.toRawHandler = function(toplevelNode) {
+            var that = this;
+            var worldFunction = adaptWorldFunction(that.handler);
+            return rawJsworld.on_particle(worldFunction, that.event, that.options);
+          };
+
+
+          //////////////////////////////////////////////////////////////////////
+          
+          var ToParticle = function(handler, ename, sdict) {
+            WorldConfigOption.call(this, 'to-particle');
+            this.handler = handler;
+            this.event = ename;
+            this.options = sd_to_js(sdict);
+          };
+          
+          ToParticle.prototype = Object.create(WorldConfigOption.prototype);
+          
+          ToParticle.prototype.toRawHandler = function(toplevelNode) {
+            var that = this;
+            var worldFunction = adaptWorldFunction(that.handler);
+            var eventGen = function(w, k) {
+              worldFunction(w, function(v) {
+                if(ffi.isSome(v)) {
+                  var xhr = new XMLHttpRequest();
+                  xhr.open("POST","https://api.particle.io/v1/devices/events");
+                  xhr.setRequestHeader("Content-type","application/x-www-form-urlencoded");
+                  xhr.send("access_token=" + that.options.acc +
+                           "&name=_event&data=" + that.event + ":" +
+                           runtime.getField(v, "value") + "&private=true&ttl=60");
```

Comment:
Yeah I figured this is tricky.  I think it'll either end up being silent failure, or find some way to check for any kind of failure and just kill the world if anything at all fails, so the user knows something is wrong.